conf.c

```diff
@@ -93,168 +128,482 @@ chomp(char *str)
-			*strchr(line, '#') = 0;
+			/* Call the check function, if it exists */
+			if(item->checkFunction != NULL)
+				if(item->checkFunction(identifier, value) != 0)
+					return EX_CONFIG;
 
-		data = line;
-		word = strsep(&data, EQS);
+			/* Else let's store the value and return */
+			if(item->str_value != NULL)
+				free(item->str_value);	/* Let's free the old value first */
 
-		/* Ignore empty lines */
-		if (word == NULL || *word == 0)
-			continue;
+			item->str_value = value;
+			return 0;
+		}
+		else	/* No value needed */
+		{
+			/* Check if a value has been provided even though no value is expected */
+			if(value != NULL)
+				return EX_CONFIG;
 
-		if (data != NULL && *data != 0)
-			data = strdup(data);
-		else
-			data = NULL;
-
-		if (strcmp(word, "SMARTHOST") == 0 && data != NULL)
-			config.smarthost = data;
-		else if (strcmp(word, "PORT") == 0 && data != NULL)
-			config.port = atoi(data);
-		else if (strcmp(word, "ALIASES") == 0 && data != NULL)
-			config.aliases = data;
-		else if (strcmp(word, "SPOOLDIR") == 0 && data != NULL)
-			config.spooldir = data;
-		else if (strcmp(word, "AUTHPATH") == 0 && data != NULL)
-			config.authpath= data;
-		else if (strcmp(word, "CERTFILE") == 0 && data != NULL)
-			config.certfile = data;
-		else if (strcmp(word, "MAILNAME") == 0 && data != NULL)
-			config.mailname = data;
-		else if (strcmp(word, "MASQUERADE") == 0 && data != NULL) {
-			char *user = NULL, *host = NULL;
-			if (strrchr(data, '@')) {
-				host = strrchr(data, '@');
-				*host = 0;
-				host++;
-				user = data;
-			} else {
-				host = data;
-			}
-			if (host && *host == 0)
-				host = NULL;
-                        if (user && *user == 0)
-                                user = NULL;
-			config.masquerade_host = host;
-			config.masquerade_user = user;
-		} else if (strcmp(word, "STARTTLS") == 0 && data == NULL)
-			config.features |= STARTTLS;
-		else if (strcmp(word, "FINGERPRINT") == 0) {
-			if (strlen(data) != SHA256_DIGEST_LENGTH * 2) {
-				errlogx(EX_CONFIG, "invalid sha256 fingerprint length");
-			}
-			unsigned char *fingerprint = malloc(SHA256_DIGEST_LENGTH);
-			if (fingerprint == NULL) {
-				errlogx(EX_CONFIG, "fingerprint allocation failed");
-			}
-			for (unsigned int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
-				if(sscanf(data + 2 * i, "%02hhx", &fingerprint[i]) != 1) {
-					errlogx(EX_CONFIG, "failed to read fingerprint");
+			if(item->checkFunction != NULL)
+				if(item->checkFunction(identifier, value) != 0)
+					return EX_CONFIG;
+
+			/* else let's set the config flag to on
+			 * we do not bother changing the value if it is already set,
+			 * as its concrete value is meaningless.
+			 * The value can already be set,
+			 * if its flag has been enabled more than once in the config file. */
+			if(item->str_value == NULL)
+				item->str_value = strdup("ON");
+			return 0;
+		}
+	}
+	else /* Item has not been found */
+		return EX_CONFIG;
+}
+
+void free_all_configuration_settings(void)
```

Comment:
the free_all_* functions are not needed, because we don't free on exit.

Reply:
Okay, I'll remove them.